### PR TITLE
Top Down Checkpoint

### PR DESCRIPTION
**Pull request by cryptoAtwill**
This opens a series of PRs on integrating IPC Top Down Checkpoint mechanism into fendermint.

This PR contains:
- Sequential key value cache that requires the key to be numeric and sequential. This is for caching block hashes using block height as key, which requires the block heights to be sequentially stored. It is using `VecDeque` as the underlying data structure to enforce sequential. Also enforcing sequential makes parent reorg detection easier.

#### Follow Up PRs:
- https://github.com/consensus-shipyard/fendermint/pull/205
- https://github.com/consensus-shipyard/fendermint/pull/206

### CODE REVIEW

File: fendermint/vm/topdown/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+mod cache;
+
+use ipc_sdk::cross::CrossMsg;
+use ipc_sdk::ValidatorSet;
+use serde::{Deserialize, Serialize};
+
+type BlockHeight = u64;
+
+/// The finality proof for IPC parent at certain height.
```
Comment (aakoshh) [inline fendermint/vm/topdown/src/lib.rs:12]:
```suggestion
/// The finality view for IPC parent at certain height.
```
I don't know what's the best phrase but I always find it strange to read that this structure _proves_ anything.

**Review by aakoshh — COMMENTED** [review]
This looks fine, only minor comments.

File: fendermint/vm/topdown/src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+            break;
+        }
+    }
+
+    /// Insert the key and value pair only if the key is upper_bound + 1
+    pub fn insert(&mut self, key: K, val: V) -> SequentialCacheInsert {
+        if let Some(upper) = self.upper_bound() {
+            if upper.add(self.increment) == key {
+                self.data.push_back((key, val));
+                return SequentialCacheInsert::Ok;
```
[the remark this answers is not in the document]
Comment (cryptoAtwill) [reply fendermint/vm/topdown/src/cache.rs:112]: I have updated the code struct, should be more readable now.

File: fendermint/vm/topdown/src/cache.rs [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    fn range_works() {
+        let mut cache = SequentialKeyCache::new(1);
+
+        for k in 0..100 {
+            cache.insert(k, k);
+        }
+
+        let range = cache.values_from(50);
+        assert_eq!(
+            range.into_iter().cloned().collect::<Vec<_>>(),
```
[the remark this answers is not in the document]
Comment (cryptoAtwill) [reply fendermint/vm/topdown/src/cache.rs:161]: Updated, nice suggestion!

**Review by adlrocha — COMMENTED** [review]
Overall LGTM. I have a slight concern on how and for what will the cache be used, but I guess I will get more context in the follow-up PRs. If this is not blocking you, I will defer approving the PR to after I review those :)

**Review by adlrocha — APPROVED** [review]
As expected this had more sense after reviewing the follow-up PR :) Thanks!

**Pull request closed without merging**